sympy/physics/control/lti.py

```diff
@@ -1259,3 +1457,509 @@ def _eval_rewrite_as_TransferFunction(self, num, den, **kwargs):
+
+    These unevaluated ``Series`` or ``Parallel`` objects can convert into the
+    resultant transfer function matrix using ``.doit()`` method or by
+    ``.rewrite(TransferFunctionMatrix)``.
+
+    >>> (-TFM8 + TFM10 + TFM8*TFM9).doit()
+    TransferFunctionMatrix([Parallel(TransferFunction(-3, s + 2, s), TransferFunction(a + s, s**2 + s + 1, s), Series(TransferFunction(3, s + 2, s), TransferFunction(-3, s + 2, s))), Parallel(TransferFunction(-p**4 + 3*p - 2, p + s, s), TransferFunction(p**4 - 3*p + 2, p + s, s), Series(TransferFunction(p**4 - 3*p + 2, p + s, s), TransferFunction(-3, s + 2, s))), Parallel(TransferFunction(a + s, s**2 + s + 1, s), TransferFunction(-a + p, 9*s - 9, s), Series(TransferFunction(-a - s, s**2 + s + 1, s), TransferFunction(-3, s + 2, s)))])
+    >>> (-TFM7 * -TFM8 * -TFM9).rewrite(TransferFunctionMatrix)
+    TransferFunctionMatrix([Series(Parallel(Series(TransferFunction(a - p, 9*s - 9, s), TransferFunction(-3, s + 2, s)), Series(TransferFunction(a + s, s**2 + s + 1, s), TransferFunction(-p**4 + 3*p - 2, p + s, s)), Series(TransferFunction(-3, s + 2, s), TransferFunction(a + s, s**2 + s + 1, s))), TransferFunction(3, s + 2, s)), Series(Parallel(Series(TransferFunction(p**4 - 3*p + 2, p + s, s), TransferFunction(-3, s + 2, s)), Series(TransferFunction(-a + p, 9*s - 9, s), TransferFunction(-p**4 + 3*p - 2, p + s, s)), Series(TransferFunction(3, s + 2, s), TransferFunction(a + s, s**2 + s + 1, s))), TransferFunction(3, s + 2, s))])
+
+    And ``.doit().doit()`` on this resultant transfer function matrix will expand the internal ``Series``
+    or ``Parallel`` objects.
+
+    >>> (-TFM8 + TFM10 + TFM8*TFM9).doit().doit()
+    TransferFunctionMatrix([TransferFunction((s + 2)**2*(-3*s**2 - 3*s + (a + s)*(s + 2) - 3) - 9*(s + 2)*(s**2 + s + 1), (s + 2)**3*(s**2 + s + 1), s), TransferFunction((p + s)*(-3*p**4 + 9*p - 6), (p + s)**2*(s + 2), s), TransferFunction((3*a + 3*s)*(9*s - 9)*(s**2 + s + 1) + (s + 2)*((-a + p)*(s**2 + s + 1) + (a + s)*(9*s - 9))*(s**2 + s + 1), (s + 2)*(9*s - 9)*(s**2 + s + 1)**2, s)])
+
+    See Also
+    ========
+
+    TransferFunction, Series, Parallel, Feedback
+
+    """
+    def __new__(cls, arg, shape=None, var=None):
+        if not (isinstance(arg, (list, tuple)) and
+            (all(isinstance(i, (list, tuple)) for i in arg) or
+            all(isinstance(i, (TransferFunction, Parallel, Series)) for i in arg))):
+            raise TypeError("Unsupported type for argument of TransferFunctionMatrix.")
+
+        if shape and var:
+            if not isinstance(shape, tuple):
+                raise TypeError("Shape must be a tuple, not {}.".format(type(shape)))
+            if not isinstance(var, Symbol):
+                raise TypeError("Var must be a Symbol, not {}.".format(type(var)))
+
+        obj = super(TransferFunctionMatrix, cls).__new__(cls, arg)
```

Comment:
I'm starting to think you really do want to reuse the matrix classes here. So I'd propose `__new__` looks like:
```python
def __new__(cls, arg, shape, var):
    if var is not None and not isinstance(var, Symbol):
        raise TypeError("Var must be a Symbol, not {}.".format(type(var)))
    if shape is not None and not isinstance(shape, tuple):
        raise TypeError("Shape must be a tuple, not {}.".format(type(shape)))

    # let matrix handle conversion to list
    matrix = ImmutableDenseMatrix(arg)
    for entry in matrix.args[2]:
        if not isinstance(entry, (TransferFunction, Parallel, Series)):
            raise TypeError("Unsupported type for argument of TransferFunctionMatrix.")
        if var is not None and entry.var != var:
            raise ValueError("All transfer functions should use the same complex"
                        " variable (var) of the Laplace transform.")
    
    if shape is not None and matrix.shape != shape:
        raise ValueError("wrong shape")  # or a better message

    obj = super(TransferFunctionMatrix, cls).__new__(cls, matrix)
```

A big advantage of this approach is that `TransferFunctionMatrix.subs(some_variable, 0)` will just work.

Reply:
Does `subs` not work with this class?

There is a basic problem in sympy with allowing non-Expr (even non-Basic) instances in places such as Matrix entries that are clearly written for Expr instances. I intend to eliminate all places where non-Expr instances are used in a Matrix from the codebase and deprecate having non-Expr in a Matrix.

I don't think that it is a big deal to implement something like matrix multiplication which is fairly trivial although perhaps it would be better if there was a reusable function somewhere for doing that calculation. It is still a work in progress and not at all documented but I have implemented various matrix routines in such a way that they can be applied to a standard Python list of lists based on only assuming that the elements support the field operations `+`, `-`, `*` and `/` so do not need to Expr or even Basic. My intention was that it should not be necessary to subclass Matrix and insert objects that don't belong there just to reuse the routines for standard matrix operations. Matrix multiplication is handled by this function:
https://github.com/sympy/sympy/blob/500f3c9a75020adc7239b50d4deac6144f342396/sympy/polys/domainmatrix.py#L220-L226
That implementation is optimised for large matrices and cheap elementary arithmetic (e.g. `int` or `mpz` as the elements).

Reply:
> Does subs not work with this class?

My impression was that `subs` would not recurse into list arguments, but perhaps I missed a sympify somewhere.

At any rate, my suggestion above is more about letting the matrix constructor handle the normalization, and less about the multiplication itself. The thing I don't like about this PR is how the special cases for `TFM([tf])` vs `TFM([[tf]])` are in every single method, when they should just be in the constructor.

Reply:
> I intend to eliminate all places where non-Expr instances are used in a Matrix from the codebase and deprecate having non-Expr in a Matrix.

If you do this, I'd argue that it might be worth introducing a base class that works over non-exprs too.

Reply:
> I'd argue that it might be worth introducing a base class

The Matrix class is already arranged into a bunch of different base classes:
```julia
In [1]: Matrix.mro()                                                                                                                                          
Out[1]: 
[sympy.matrices.dense.MutableDenseMatrix,
 sympy.matrices.dense.DenseMatrix,
 sympy.matrices.matrices.MatrixBase,
 sympy.matrices.matrices.MatrixDeprecated,
 sympy.matrices.matrices.MatrixCalculus,
 sympy.matrices.matrices.MatrixEigen,
 sympy.matrices.matrices.MatrixSubspaces,
 sympy.matrices.matrices.MatrixReductions,
 sympy.matrices.matrices.MatrixDeterminant,
 sympy.matrices.common.MatrixCommon,
 sympy.matrices.common.MatrixArithmetic,
 sympy.matrices.common.MatrixOperations,
 sympy.matrices.common.MatrixProperties,
 sympy.matrices.common.MatrixSpecial,
 sympy.matrices.common.MatrixShaping,
 sympy.matrices.common.MatrixRequired,
 sympy.printing.defaults.Printable,
 object]
```
It's just not really clear how you could make use of any of those classes separately.